src/components/syncplay/syncplayManager.js

```diff
@@ -0,0 +1,737 @@
+
+        events.on(timeSyncManager, "Update", (event, timeOffset, ping) => {
+            this.timeOffsetWithServer = timeOffset;
+            this.roundTripDuration = ping * 2;
+
+            if (this.notifySyncplayReady) {
+                this.syncplayReady = true;
+                events.trigger(this, "SyncplayReady");
+                this.notifySyncplayReady = false;
+            }
+
+            // Report ping
+            if (this.syncEnabled) {
+                const apiClient = connectionManager.currentApiClient();
+                const sessionId = getActivePlayerId();
+
+                if (!sessionId) {
+                    this.signalError();
+                    toast({
+                        // TODO: translate
+                        text: "Syncplay error occured."
+                    });
+                    return;
+                }
+
+                apiClient.sendSyncplayCommand(sessionId, "UpdatePing", {
+                    Ping: ping
+                });
+            }
+        });
+    }
+
+    /**
+     * Called when the player changes.
+     */
+    onPlayerChange () {
+        this.bindToPlayer(playbackManager.getCurrentPlayer());
+        events.trigger(this, "PlayerChange", [this.currentPlayer]);
+    }
+
+    /**
+     * Called on playback state changes.
+     * @param {Object} e The playback state change event.
+     */
+    onPlayPauseStateChanged (e) {
+        events.trigger(this, "PlayPauseStateChange", [this.currentPlayer]);
+    }
+
+    /**
+     * Called on playback progress.
+     * @param {Object} e The time update event.
+     */
+    onTimeUpdate (e) {
+        // NOTICE: this event is unreliable, at least in Safari
+        // which just stops firing the event after a while.
+        events.trigger(this, "TimeUpdate", [e]);
+    }
+
+    /**
+     * Called when playback is resumed.
+     */
+    onPlaying () {
+        // TODO: implement group wait
+        this.lastPlaybackWaiting = null;
+        events.trigger(this, "PlayerPlaying");
+    }
+
+    /**
+     * Called when playback is buffering.
+     */
+    onWaiting () {
+        // TODO: implement group wait
+        if (!this.lastPlaybackWaiting) {
+            this.lastPlaybackWaiting = new Date();
+        }
+        events.trigger(this, "PlayerWaiting");
+    }
+
+    /**
+     * Gets playback buffering status.
+     * @returns {boolean} _true_ if player is buffering, _false_ otherwise.
+     */
+    isBuffering () {
+        if (this.lastPlaybackWaiting === null) return false;
+        return (new Date() - this.lastPlaybackWaiting) > this.minBufferingThresholdMillis;
+    }
+
+    /**
+     * Binds to the player's events.
+     * @param {Object} player The player.
+     */
+    bindToPlayer (player) {
+        if (player !== this.currentPlayer) {
+            this.releaseCurrentPlayer();
+            this.currentPlayer = player;
+            if (!player) return;
+        }
+        
+        // TODO: remove this extra functions
```

Comment:
Because the `events` module changes the scope when calling a callback.
Calling for instance `onPlayPauseStateChanged` from the wrong scope breaks things, because `this` points to `player` (the event emitter) instead of pointing to the `SyncplayManager` singleton.

This was a quick fix to avoid spending to much time on it. Is there a cleaner way or should the `events` module be rethinked?

Reply:
Please add this explanation a `//FIXME` comment